Render annotation flags over track group headers

cdamus opened the pull request:
All track rendering was done on a big scrolling canvas element that sits behind the track dividing lines and shells. That is fine for rendering stuff in the track content area of the scrolling tracks, but the track content area of the headers of expanded groups does not scroll with the rest. They're not even a part of the same static layout, instead being positioned relatively and moreover sticking to the top of the panel as tracks scroll up beneath them. Because they are in a different layout to the canvas, they occlude anything painted on the canvas, so painting in the expanded group headers has no effect.

Consequently, it is necessary to give each expanded group header its own canvas on which we can draw content such as the stems of note flags.

![CleanShot 2024-02-28 at 12 57 14](https://github.com/android-graphics/sokatoa/assets/1615558/1f3d0de6-054f-4ca7-91c8-7ce8b9dcf284)

cdamus committed:
Render annotation flags over track group headers

All track rendering was done on a big scrolling canvas element that sits
behind the track dividing lines and shells. That is fine for rendering
stuff in the track content area of the scrolling tracks, but the track
content area of the headers of expanded groups does not scroll with the
rest. They're not even a part of the same static layout, instead being
positioned relatively and moreover sticking to the top of the panel as
tracks scroll up beneath them. Because they are in a different layout
to the canvas, they occlude anything painted on the canvas, so painting
in the expanded group headers has no effect.

Consequently, it is necessary to give each expanded group header its own
canvas on which we can draw content such as the stems of note flags.

Signed-off-by: Christian W. Damus <cdamus.ext@eclipsesource.com>

ALevansSamsung commented:
Adding or removing a note while the sticky open track group is still there tends to make the annotations drawn on it stale.

https://github.com/eclipsesource/perfetto/assets/121060410/81184ce8-7b9c-4e77-9157-44986432d073

cdamus committed:
Fix chost annotation legs persisting after deletion

The algorithm for determining what needs to be redrawn didn't account
for the fact that a group header that would be scrolled offscreen and
so not need redrawing when collapsed might be sticky and still onscreen
when expanded. So now we redraw an expanded group header if it really
seems to be onscreen _or_ if any member within it (recursively) is
onscreen (which would imply that the group header is sticky at the top.)

Signed-off-by: Christian W. Damus <cdamus.ext@eclipsesource.com>

cdamus commented:
> Adding or removing a note while the sticky open track group is still there tends to make the annotations drawn on it stale.

Great find! Thanks. That was a tricky case and should be fixed in commit 4a6af55314e9f41d222254e2d27f5d70263bfb93.

ALevansSamsung commented:
While the annotation lines are no longer stale, the ones added while the sticky header is up are not the same length
![image](https://github.com/eclipsesource/perfetto/assets/121060410/e8d3ae6e-c93e-4503-be68-671c13be482a)

cdamus commented:
> While the annotation lines are no longer stale, the ones added while the sticky header is up are not the same length

Ah. What's happening here is that a group header has tucked up underneath another header that is not as high. In light mode, you would see that the header on top has a line ① running along its bottom edge and the apparent gap ② is just where the header underneath doesn't know it still needs to be drawn.

![CleanShot 2024-02-29 at 08 18 10](https://github.com/eclipsesource/perfetto/assets/1615558/1c78ce79-045b-4717-9b1b-5e8c9646a77c)

cdamus committed:
Just always paint sticky panels in bottom-up order

Signed-off-by: Christian W. Damus <cdamus.ext@eclipsesource.com>

cdamus commented:
> While the annotation lines are no longer stale, the ones added while the sticky header is up are not the same length

@ALevansSamsung this should be fixed with commit b450b905b677b68128418d9e0692744e58a959a1 which, overall, simplifies the code greatly by not trying to be clever about efficient drawing of something is known not to be complex to draw 😀

ALevansSamsung commented:
Still seems to be an issue with staleness
https://github.com/eclipsesource/perfetto/assets/121060410/41f7d461-7cb5-4413-9d4b-385108d611e1
